### PR TITLE
Release 0.28.0

### PR DESCRIPTION
**Pull request by apoelstra**
See discussion here https://github.com/rust-bitcoin/rust-bitcoin/discussions/941 about whether we want to cut the release now or do another slew of breaking changes between RCs.

### CODE REVIEW

**Review by sanket1729 — APPROVED** [review]
Wow, thanks for making the release notes with a good description. ACK 785be0e27a419955bd83ff5fd599c38852a3ce2c

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,63 @@
...
+it is also **by far** our largest release ever, at 148 PRs merged from 23
+different contributors. Our primary goal in this release was to introduce
+support for Taproot and its associated data structures: addresses, taptrees,
+sighashes, PSBT fields, and more. As it turned out, these changes required
+(or at least, incentivized) changing a lot of our APIs, causing a significant
+increase in scope.
+
+We have more big changes coming down the pike. 2022 is going to be a big
+year for `rust-bitcoin`, which we know is exciting for us but disruptive to
+downstream users who ultimately want the library to just wark. Our hope is
```
Comment (RCasatta) [inline CHANGELOG.md:14]: *work

**Comment by dr-orlovsky** [comment]
@apoelstra thank you for this wonderful change log, it looks like a huge work nobody was willing to do!

ACK 785be0e27a419955bd83ff5fd599c38852a3ce2c

We have just two PRs left with RC fixes (quite important bug in PSBT taptree with hidden nodes #929 + trivial but breaking API changes to method names #927). They are reviewed multiple times. If you will have time to review we may get two ACKs today and merge pre-release. Otherwise feel free to release w/o them.

**Review by dr-orlovsky — APPROVED** [review]
ACK 785be0e27a419955bd83ff5fd599c38852a3ce2c

**Comment by sanket1729** [comment]
I have ACKs for both #929 and #924. None of them need to be mentioned in release notes(I think). 

I think we should try to get #929 in if possible.
#924 is good to have, but not really needed.

**Comment by dr-orlovsky** [comment]
> None of them need to be mentioned in release notes(I think).

Agree: they are just fixes to the previous PRs

**Comment by apoelstra** [comment]
Ok I'll try to get 929 and 924 in tonight or tomorrow, before release.

File: CHANGELOG.md (hunk 1 above)
Comment (apoelstra) [reply CHANGELOG.md:14]: Will fix :) thanks

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,63 @@
...
+of the individual crates in this ecosystem while still providing the essential
+functionality needed by our downstream users, especially wallet projects.
+
+We will also develop a plan to make our releases more predictable and manageable,
+likely by having scheduled releases with limited scope. We would like to reach
+a point where we no longer have frequent breaking releases, but right now we
+are nowhere close.
+
+Upcoming changes will include
+- A quick new release which updates our MRSV from 1.29 to 1.41 and does little else
```
Comment (apoelstra) [inline CHANGELOG.md:26]: @Kixunil I know you would like to get some more breaking changes in related to Taproot and PSBT. Would it be ok if we bumped the MSRV first?

**Comment by dr-orlovsky** [comment]
@apoelstra BTW I already used the current release downstream in my projects and tested quite advanced taproot scenarious, including full OP_RETURN taproot commitment implementation and complex miniscript-based script-path non-SIGHASH_ALL spendings with mutlsigs and timelocks on mainnet. Everything works fine; and the API is finally (after all RC fixes) covers all what is needed.

**Comment by tcharding** [comment]
BOOM! 

ACK [785be0e](https://github.com/rust-bitcoin/rust-bitcoin/pull/944/commits/785be0e27a419955bd83ff5fd599c38852a3ce2c)

**Review by tcharding — APPROVED** [review]

**Review by dr-orlovsky — DISMISSED** [review]
tACK: cb4d34fd400d2c6ffefc140e9d7f06c5b3e82b0d

spent taproot output on a script path spent which contained 1-of-3 `multi_a` + timelock https://blockstream.info/testnet/tx/62c5c416a5b26c1e4e6c30dff6497075f332b797c22ffa98e837f2e74a64c569?expand using current master at cb4d34fd400d2c6ffefc140e9d7f06c5b3e82b0d + rust-miniscript at 909f0a51c5881aafd731386a46d4756b30ee2c79 (from https://github.com/rust-bitcoin/rust-miniscript/pull/337)

File: CHANGELOG.md (hunk 2 above)
Comment (Kixunil) [reply CHANGELOG.md:26]: I don't remember which Taproot/PSBT breaking changes I wanted but yes, I do think bumping MSRV should be one of the first tasks. That being said, I suggest we try to merge as many open PRs first to minimize the churn. I'd rather redo semi-automated edition fixes than rebase all other PRs.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,63 @@
...
+- Change BIP32 to use rust-secp256k1 keys rather than rust-bitcoin ones (no compressedness flag) [#590](https://github.com/rust-bitcoin/rust-bitcoin/pull/590) [#591](https://github.com/rust-bitcoin/rust-bitcoin/pull/591)
+- Rename inner key field in `PrivateKey` and `PublicKey` [#762](https://github.com/rust-bitcoin/rust-bitcoin/pull/762)
+- Address and denomination related changes [#768](https://github.com/rust-bitcoin/rust-bitcoin/pull/768) [#784](https://github.com/rust-bitcoin/rust-bitcoin/pull/784)
+- Don't allow hybrid EC keys [#829](https://github.com/rust-bitcoin/rust-bitcoin/pull/829)
+- Change erroneous behavior for `SIGHASH_SINGLE` bug [#860](https://github.com/rust-bitcoin/rust-bitcoin/pull/860) [#897](https://github.com/rust-bitcoin/rust-bitcoin/pull/897)
+- Delete the deprecated `contracthash` module [#871](https://github.com/rust-bitcoin/rust-bitcoin/pull/871); this functionality will migrate to ElementsProject/rust-elements
+- Remove compilation-breaking feature-gated enum variants [#881](https://github.com/rust-bitcoin/rust-bitcoin/pull/881)
```
Comment (Kixunil) [inline CHANGELOG.md:55]: The wording of this sounds like the variants were removed while actually they are always present. Perhaps something like "Feature-gated enum variants are now always present but use placeholder types when features are disabled."

Comment (apoelstra) [reply CHANGELOG.md:55]: How about "Remove compilation-breaking feature-gating of enum variants"?

Comment (Kixunil) [reply CHANGELOG.md:55]: A bit simplistic (there's still feature gating but done differently) but sounds acceptable to me. :)

**Review by sanket1729 — APPROVED** [review]
utACK 38f09d2d936f923d16bda5fc0214590ac0f9e9e5. The diff from the previous ACKs is minor easy to review.

**Review by tcharding — APPROVED** [review]
Lets go!

**Comment by Kixunil** [comment]
Can we ACK #927 before releasing? It's super simple and quite nice.

**Comment by apoelstra** [comment]
Merged 927. Merged this. Tagged and publish. We have liftoff!

**Pull request merged**